src/dictarray.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2009-2023, Redis Ltd.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *   * Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of Redis nor the names of its contributors may be used
+ *     to endorse or promote products derived from this software without
+ *     specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DICTARRAY_H_
+#define DICTARRAY_H_
+
+#include "dict.h"
+#include "adlist.h"
+
+typedef struct {
+    list *rehashing;                       /* List of dictionaries in this dictarray that are currently rehashing. */
+    int resize_cursor;                     /* Cron job uses this cursor to gradually resize dictionaries. */
+    int non_empty_slots;                   /* The number of non-empty slots. */
+    unsigned long long key_count;          /* Total number of keys in this dictarray. */
+    unsigned long long bucket_count;       /* Total number of buckets in this dictarray across dictionaries. */
+    unsigned long long *slot_size_index;   /* Binary indexed tree (BIT) that describes cumulative key frequencies up until given slot. */
+} daState;
+
+typedef struct {
+    dict **dicts;
+    long long num_slots;
```

Comment:
i wonder if the term "slots" is right in that context?
maybe it should be dicts, and dictIndex?
i.e. the hashing and slot indexing is still calculated from outside (db.c), right?
so the dict-array is just a generic array of dictionaries, nothing specific to hash-slots?
and the dbXxxx interfaces are the ones that do the slot magic?

Reply:
yes, i guess i can just call it `length` and `length_bits`?

Reply:
we need to consider all the other references to "slot" and "slots" in this file.
it could be "length" / "count" and "index". or "num_dicts" and "dict_idx".
i suppose the right way to find out is to try and rename them all and see what feels right.

Reply:
Yeah, we discussed this in the per-slot dictionary PR, but it didn't really fit there.  I agree, we should just rename stuff and see what sticks.